specs/granite/header_accumulator.md

```diff
@@ -0,0 +1,302 @@
+| `TOTAL_TREE_DEPTH`        | `ACCUMULATOR_TREE_DEPTH + HEADER_BATCH_TREE_DEPTH`                                                                                         |
+| `TRUNCATED_COMMITMENT`    | `20` bytes. Must be sufficiently long to protect against hash collision.                                                                   |
+| `MERKLE_STACK_SIZE`       | `TRUNCATED_COMMITMENT * ACCUMULATOR_TREE_DEPTH`                                                                                            |
+| `HEADER_BATCH_SIZE`       | `2 ** HEADER_BATCH_TREE_DEPTH`                                                                                                             |
+| Accumulator Tree          | A partial merkle tree of depth `ACCUMULATOR_TREE_DEPTH`, storing the root nodes of header batch trees as leaves.                           |
+| Header Batch Tree         | A binary merkle tree of depth `HEADER_BATCH_TREE_DEPTH`, storing the header hashes of `HEADER_BATCH_SIZE` block headers as leaves.         |
+| Header Hash               | `keccak256(rlp(header))`                                                                                                                   |
+| Merkle Stack              | The path to the latest node added to the Accumulator Tree. This is an array of `MERKLE_STACK_SIZE` bytes in length.                        |
+| Partial Merkle Tree       | As specified in [_End-to-End Formal Verification of Ethereum 2.0 Deposit Smart Contract_][imt] by Daejun Park, Yi Zhang, and Grigore Rosu. |
+| Zero Hashes               | As specified in [_End-to-End Formal Verification of Ethereum 2.0 Deposit Smart Contract_][imt] by Daejun Park, Yi Zhang, and Grigore Rosu. |
+
+## Accumulator Tree Construction
+
+The header accumulator tree is constructed as a partial merkle tree of depth `ACCUMULATOR_TREE_DEPTH`, with leaves
+representing the merkle roots of binary header batch trees, containing `HEADER_BATCH_SIZE` blocks.
+
+As an example, below is an accumulator tree with `ACCUMULATOR_TREE_DEPTH = 1` and `HEADER_BATCH_TREE_DEPTH = 2`,
+holding a total of `8` blocks' header hashes across two batches:
+
+```mermaid
+flowchart TD
+  A[Header Accumulator Root] --> B[Header Batch #1]
+
+  A --> C[Header Batch #2]
+
+  B --> B1[Intermediate]
+  B --> B2[Intermediate]
+  B1 --> B11[Header Hash #1]
+  B1 --> B12[Header Hash #2]
+  B2 --> B21[Header Hash #3]
+  B2 --> B22[Header Hash #4]
+
+  C --> C1[Intermediate]
+  C --> C2[Intermediate]
+  C1 --> C11[Header Hash #5]
+  C1 --> C12[Header Hash #6]
+  C2 --> C21[Header Hash #7]
+  C2 --> C22[Header Hash #8]
+```
+
+Both the global accumulator tree as well as the header batch tree use `keccak256` as their hashing function. For all
+nodes within the header accumulator tree, all commitments are shortened to `TRUNCATED_COMMITMENT` bytes in length to
```

Comment:
One thing to note is that if we only have a truncated commitment, we can't use it to retrieve data from the preimage oracle - we need the full hash for that.

Reply:
We can add a new `PreimageKeyType` for this, yes? Should be the same as the `keccak` route, just need to truncate the resulting digest.

Reply:
Yeh I think that should work.  This is where I think having those details fleshed out would be really helpful - we don't want to implement this spec and then find there's something that's suboptimal for actually using it in fault proofs, so it's worth making sure we're actually thinking through all the details of how it will be used.